rtgui/colortoning.cc

```diff
@@ -1195,3 +1372,195 @@ bool ColorToning::resetPressed(GdkEventButton* event)
     return false;
 }
+
+
+bool ColorToning::labRegionResetPressed(GdkEventButton *event)
+{
+    labRegionAB->reset(event->state & GDK_CONTROL_MASK);
+    return false;
+}
+
+
+void ColorToning::onLabRegionSelectionChanged()
+{
+    auto s = labRegionList->get_selected();
+    if (!s.empty()) {
+        // update the selected values
+        labRegionGet(labRegionSelected);
+        labRegionSelected = s[0];
+        labRegionShow(labRegionSelected);
+        if (labRegionShowMask->get_active()) {
+            labRegionShowMaskChanged();
+        }
+    }
+}
+
+
+void ColorToning::labRegionGet(int idx)
+{
+    auto &r = labRegionData[idx];
+    double la, lb;
+    labRegionAB->getParams(la, lb, r.a, r.b);
+    r.saturation = labRegionSaturation->getValue();
+    r.lightness = labRegionLightness->getValue();
+    r.hueMask = labRegionHueMask->getCurve();
+    r.chromaticityMask = labRegionChromaticityMask->getCurve();
+    r.lightnessMask = labRegionLightnessMask->getCurve();
+}
+
+
+void ColorToning::labRegionAddPressed()
+{
+    labRegionData.insert(labRegionData.begin(), rtengine::ColorToningParams::LabCorrectionRegion());
+    labRegionSelected = 0;
+    labRegionPopulateList();
+    labRegionShow(labRegionSelected);
+
+    if (listener) {
+        listener->panelChanged(EvLabRegionList, M("HISTORY_CHANGED"));
+    }
+}
+
+
+void ColorToning::labRegionRemovePressed()
+{
+    if (labRegionList->size() > 1) {
+        labRegionData.erase(labRegionData.begin() + labRegionSelected);
+        labRegionSelected = LIM(labRegionSelected-1, 0, int(labRegionData.size()-1));
+        labRegionPopulateList();
+        labRegionShow(labRegionSelected);
+
+        if (listener) {
+            listener->panelChanged(EvLabRegionList, M("HISTORY_CHANGED"));
+        }
+    }
+}
+
+
+void ColorToning::labRegionUpPressed()
+{
+    if (labRegionSelected > 0) {
+        auto r = labRegionData[labRegionSelected];
+        labRegionData.erase(labRegionData.begin() + labRegionSelected);
+        --labRegionSelected;
+        labRegionData.insert(labRegionData.begin() + labRegionSelected, r);
+        labRegionPopulateList();
+
+        if (listener) {
+            listener->panelChanged(EvLabRegionList, M("HISTORY_CHANGED"));
+        }
+    }
+}
+
+
+void ColorToning::labRegionDownPressed()
+{
+    if (labRegionSelected < int(labRegionData.size()-1)) {
+        auto r = labRegionData[labRegionSelected];
+        labRegionData.erase(labRegionData.begin() + labRegionSelected);
+        ++labRegionSelected;
+        labRegionData.insert(labRegionData.begin() + labRegionSelected, r);
+        labRegionPopulateList();
+
+        if (listener) {
+            listener->panelChanged(EvLabRegionList, M("HISTORY_CHANGED"));
+        }
+    }
+}
+
+
+void ColorToning::labRegionShowMaskChanged()
+{
+    if (listener) {
+        listener->panelChanged(EvLabRegionShowMask, labRegionShowMask->get_active() ? M("GENERAL_ENABLED") : M("GENERAL_DISABLED"));
+    }
+}
+
+
+namespace {
```

Comment:
Please move the anonymous namespace with helpers to the top of the file, so that there's ideally only one such namespace. This is where everyone will look for it when adding other helpers.